manifests/plugin/container_gateway.pp

```diff
@@ -22,13 +36,34 @@
   Boolean $enabled = true,
   Foreman_proxy::ListenOn $listen_on = 'https',
   Stdlib::HTTPUrl $pulp_endpoint = "https://${facts['networking']['fqdn']}",
+  String $database_backend = 'postgres',
   Stdlib::Absolutepath $sqlite_db_path = '/var/lib/foreman-proxy/smart_proxy_container_gateway.db',
   Optional[Integer] $sqlite_timeout = undef,
+  Boolean $manage_postgresql = true,
+  Stdlib::Host $postgresql_host = 'localhost',
+  Stdlib::Port $postgresql_port = 5432,
+  String $postgresql_database = 'container_gateway',
+  String $postgresql_user = 'foreman-proxy',
+  String $postgresql_password = extlib::cache_data('container_gateway_cache_data', 'db_password', extlib::random_password(32))
```

Comment:
AFAIK the installer doesn't support a function as a default value. The classic way we've dealt with this is introducing `$class::params::$param`. An alternative that I've been contemplating is making the value `undef` and then determine it in the body using [`lest`](https://www.puppet.com/docs/puppet/8/function.html#lest):
```puppet
class mycls (
  Optional[String] $param = undef,
) {
  $_param = $param.lest || { extlib::cache_data('container_gateway_cache_data', 'db_password', extlib::random_password(32)) }
}
```

This has the benefit that the password doesn't show up in the answers file and if a value is provided, it doesn't write down an unused cache file with some random password.

Again, something we don't apply anywhere else but something I've been thinking about.

Then another matter is [`Sensitive`](https://www.puppet.com/docs/puppet/8/securing-sensitive-data.html#securing_sensitive_data-epp-templates), which we barely use right now but it automatically does the right thing like hide the output of it in diffs (which users have complained about in the past). It does mean the template should be changed from ERB to EPP.

Reply:
> AFAIK the installer doesn't support a function as a default value. The classic way we've dealt with this is introducing `$class::params::$param`. An alternative that I've been contemplating is making the value `undef` and then determine it in the body using [`lest`](https://www.puppet.com/docs/puppet/8/function.html#lest):

I tested it on a smart proxy and it seems to work fine. However I'm cool with changing up the method to avoid having the password show up in the answers file.